R/shiny.R

```diff
@@ -53,7 +53,18 @@ renderDataTable = function(expr, env = parent.frame(), quoted = FALSE, ...) {
   )
   checkShinyVersion()
   if (!quoted) expr = substitute(expr)
-  htmlwidgets::shinyRenderWidget(expr, dataTableOutput, env, quoted = TRUE)
+
+  exprFunc <- shiny::exprToFunction(expr, env, quoted = TRUE)
+  widgetFunc <- function() {
+    x <- exprFunc()
+    if (is.data.frame(x)) {
+      datatable(x)
+    } else {
+      x
+    }
```

Comment:
Actually I think we could use this spot in the code to implement `server=TRUE` without requiring help from htmlwidgets. This is effectively the same thing as a Shiny-only preRenderHook, and the name and session are both available I believe.